Genesis run by init, terminals can tail the log

nshalman opened the pull request:
Fixes #66 

With this change applied, genesis kindly puts shells on the serial consoles.

roymarantz commented on bootcd/genesis.ks.template:
```diff
@@ -220,6 +220,8 @@ echo ">>>>> Removing bloat from /usr/share"
 rm -rf /usr/share/doc
 rm -rf /usr/share/icons
 
+sed -e 's|ACTIVE_CONSOLES.*|ACTIVE_CONSOLES="/dev/tty1 /dev/ttyS0 /dev/ttyS1"|' -i /etc/sysconfig/init
```
Why restrict it to only /dev/tty1 instead of /dev/tty[1-6] which is the default?

nshalman replied:
IIRC the content of that file on a freshly built Genesis image without my change was just `/dev/tty1`. I have no objection to updating this to look more like `/dev/tty[1-6] /dev/ttyS[0-1]` if that would be preferable to others and also works.

roymarantz replied:
How about just appending to the line and leaving the rest alone?

nshalman replied:
I'm fairly certain the line as is doesn't contain a `"` mark after the equal sign which is needed for it to properly capture this information which has to be separated by spaces.
IIRC (it will take a few minutes for me to rebuild and check) I'm going from:
`ACTIVE_CONSOLES=/dev/tty1`
to
`ACTIVE_CONSOLES="/dev/tty[1-6] /dev/ttyS[0-1]"`
(which I just tested and does appear to work so I'll be updating the PR shortly)

"Appending" could be implemented as the slightly more complicated:
1. remove any `"` after the `=` and at the end of the line if they are there.
2. append ` /dev/ttyS[0-1]`
3. put back in the necessary quotation marks

I'm happy to spend a few minutes writing that if it would be preferable, though if the current value is only `/dev/tty1` then explicitly setting it to `/dev/tty[1-6]` could be preferable based on your original feedback. Thoughts?

nshalman replied:
As a note of clarification I've been testing with the `Dockerfile` not `Dockerfile.centos7` which may have different defaults.

nshalman replied:
Now I see https://github.com/tumblr/genesis/blob/master/bootcd/rpms/genesis_scripts/src/sysconfig-init.diff which is where it's being explicitly locked down to `tty1` per @byxorna's explanation below. I need to poke a little deeper to see if I can untangle the launching of all the automated functionality from making logging in possible.

byxorna commented:
So, this needs testing, because of some interesting decisions that were made about how to launch Genesis bootloader on start. We have an auto login script, and Genesis is launched from a .bashrc or .profile iirc, with some locking to prevent multiple concurrent runs. With this change, will all the ttys attempt to log in and run the .profile? This is undesirable imho, and we should figure out how to make only a single instance launch, not tied to a (v)tty.

nshalman commented:
@byxorna I have a bad feeling that you're correct that they all log in in a way that would be undesirable as you describe.

byxorna commented:
Yea... I've been wanting to fix how Genesis bootloader launches, so this may be a good opportunity to do so. If we can get upstart to launch it, and still get output sent to vtty1 for operator troubleshooting, that would be perfect. No more interactive login hijinks

nshalman commented:
Ideally the output should go to a runtime-configurable set of terminals. Some operators might want to watch the output over IPMI SoL or have it go to both SoL and the "VGA" so that it can be monitored from either depending on who's watching.

Perhaps the auto-login either drops to a shell, or tails a log file where the upstart launched automated run is logging?

I'll keep poking around the guts to see how much I can grok the automated bits.

byxorna commented:
Yea, I like it! Too much of Genesis isn't easily overridable​ by admins making their own spin without getting into the guts. An easy override via kernel command line and /etc file would be dope. If you can figure out an elegant way to do this, that would be awesome!!

jyundt commented:
> Ideally the output should go to a runtime-configurable set of terminals. Some operators might want to watch the output over IPMI SoL or have it go to both SoL and the "VGA" so that it can be monitored from either depending on who's watching.

Could we make this default to "all consoles"? (Assuming the tasks in .bashrc/.profile can be extracted and executed only once.)

It would be handy to have the console available "everywhere" (VGA, SOL, etc) by default instead of having to override a bootarg + reboot if the console isn't where you expect it at first.

> Yea, I like it! Too much of Genesis isn't easily overridable​ by admins making their own spin without getting into the guts. An easy override via kernel command line and /etc file would be dope.

:thumbsup:

nshalman commented:
I have a working demo of using an init script to run the genesis bootloader under `screen` and a root profile that attaches to the screen session. If that is of interest I will push my new code for you lovely folks to review.

nshalman commented on bootcd/rpms/genesis_scripts/src/run-genesis-bootloader:
```diff
@@ -0,0 +1,17 @@
+    GENESIS_PROMPT_TIMEOUT=0  /usr/bin/genesis-bootloader 2>&1 | tee /var/log/genesis-bootloader.log
+fi
+echo "Genesis done. Dropping to a shell"
+exec $SHELL
```
Still needs some work. The behavior of this line isn't what I would have expected.

nshalman commented:
Well, `screen` was heavy handed (as @jyundt mentioned to me privately.)
The init script runs a wrapper which drops a pid file and redirects stdout and stderr to the log file before running the bootloader.
All terminals log in and tail the log file until the pid obtained from the pidfile exits.

nshalman commented:
One last force push included some cleanup after I'd tested it (don't ask how many cycles of typo fixing and rebuilds I went through...)

This is looking very nice to me. Feedback requested!

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -11,6 +11,9 @@ Source2:        src/sysconfig-init.diff
 Source5:        src/login-shell
+Source6:        src/init.d-genesis
+Source7:        src/run-genesis-bootloader
+Source8:        src/sysconfig-genesis
```
Genesis.sysconfig is a better name, to line up with patterns in other rpms. (Nitpick)

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -11,6 +11,9 @@ Source2:        src/sysconfig-init.diff
 Source3:        src/tty.conf.override
 Source4:        src/genesis-bootloader
 Source5:        src/login-shell
+Source6:        src/init.d-genesis
```
(nitpick) genesis.init better name?

nshalman replied:
should I also clean up line 9 by moving and renaming the existing `init.d-network-prep`?

byxorna replied:
sure, that would be nice to have

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -25,9 +28,11 @@ Scripts and configuration files used by Genesis in the bootcd image
 mkdir -p $RPM_BUILD_ROOT/root
+# add root's bash_profile
 install -m 644 -T %{SOURCE0}   $RPM_BUILD_ROOT/root/.bash_profile.genesis_scripts
+# add bootloader wrapper to root's homedir
```
I'd prefer this to get dropped into a common bin dir instead of /root

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -49,10 +56,13 @@ install -m 555 -T %{SOURCE4}   $RPM_BUILD_ROOT/usr/bin/genesis-bootloader
+%config /etc/sysconfig/genesis
 %config /etc/init/tty.conf.override
 %config /root/.bash_profile.genesis_scripts
+%config /root/run-genesis-bootloader
```
It's not really a config though... I'd remove this config directive to prevent odd behavior on upgrade when booted into live image (i.e. when testing)

byxorna commented on bootcd/rpms/genesis_scripts/src/init.d-genesis:
```diff
@@ -0,0 +1,20 @@
+#
+
+### BEGIN INIT INFO
+# X-Start-After:    $network
```
I think we should ask for more facilities to be up. I forget the upstart (ugh) parlance, but we want syslog, all interfaces, and boot settled before proceeding.

byxorna commented on bootcd/rpms/genesis_scripts/src/root-bash_profile:
```diff
@@ -1,14 +1,25 @@
+	if [[ -f ${GENESIS_PIDFILE} ]]
+	then
+		# get the pid and break out of the for loop
+		GENESIS_PID=$(cat ${GENESIS_PIDFILE})
```
This should not be bare variable expansion. Could you quote this please?

byxorna commented on bootcd/rpms/genesis_scripts/src/root-bash_profile:
```diff
@@ -1,14 +1,25 @@
+source /etc/sysconfig/genesis
+
+# Try for two seconds to find a pidfile
+for i in {1..20}
```
This whole loop may be simplified with lsb init script functions from sourcing /lib/lsb/init-functions. There are helpers to read PIDs in a reliable manner

nshalman replied:
The filename you specify doesn't show up in my booted genesis build. Is there a package I should be pulling in to get that file (and thus those functions?)

byxorna replied:
i think its part of lsb-base or one of its sister packages. Theres also /etc/init[.d]/functions which is an overlapping set of features, you could look at instead

nshalman replied:
I'm pushing back on this one. I looked in `/etc/init.d/functions` and found nothing that does what I want which may not be totally obvious:
The pid file is used to wait for the bootloader to have started but I don't care if it's currently running.
The init script first starts logging and then drops a pid file which means if I wait for the pid file I am guaranteed that logging to the log file has begun and the tail should work correctly.

If you know of a particular function that waits a reasonable amount of time for a file to exist, I'm happy to use it, but I didn't find any obvious candidates.

byxorna commented on bootcd/rpms/genesis_scripts/src/root-bash_profile:
```diff
@@ -1,14 +1,25 @@
+
+# If we found a pid and a log file, tail the log until the
+# process exits
+if [[ ! -z ${GENESIS_PID} ]] && [[ -f ${GENESIS_LOGFILE} ]]
```
What if we didn't?! Should at least print an error. I'd posit there isn't a way we want Genesis livecd to boot without running the bootloader. Also, what happens when you run utility mode, and it just drops to shell? Will this not output anything?

nshalman replied:
because it's the bash profile, after falling off the end of that file it drops to a shell.
I can make an else clause that explains that something is wrong.

byxorna replied:
Yea, probably a warning that tells the user something is wrong and where to look for more details (log file, where genesis-bootloader bin is, etc)

byxorna commented on bootcd/rpms/genesis_scripts/src/run-genesis-bootloader:
```diff
@@ -0,0 +1,21 @@
+# send stdout and stderr to the log file
+exec &>${GENESIS_LOGFILE}
+
+echo $$>${GENESIS_PIDFILE}
```
You should assume a sane default if genesis_pidfile isn't set in the sysconfig. Same with genesis_logfile

nshalman commented:
Thanks for the feedback. I'll work on addressing your comments.
I also found some odd behavior based on what `console=` flags are or aren't passed on the kernel command line that I'm trying to clean up. There's some other simplification that will happen as a result of that cleanup.

byxorna commented:
@nshalman great, looking forward to it!

nshalman commented:
I'm pushing a bunch of individual commits that address the many pieces of feedback I have received.

I believe this should all get squashed down to a single commit before merging, but this was cleaner for me to work on while addressing the feedback.

byxorna commented on bootcd/rpms/genesis_scripts/src/root-bash_profile:
```diff
@@ -1,14 +1,33 @@
+	if [[ -f "${GENESIS_PIDFILE}" ]]
+	then
+		# get the pid and break out of the for loop
+		GENESIS_PID=$(cat "${GENESIS_PIDFILE}")
```
I was thinking something like http://refspecs.linuxbase.org/LSB_3.0.0/LSB-PDA/LSB-PDA/iniscrptfunc.html pidofproc, but no worries if you do t want to use that :)

nshalman replied:
Right, but in the case of e.g. util mode, that thing may have already exited by the time I'm asking about the pid. I just don't want to somehow try to tail the log *before* it runs.

byxorna replied:
Adding `-F` to tail will reopen the log if it's fd changes or is truncated, for example when a new process sends it's output to the file. Will this help simplify?

byxorna commented on bootcd/rpms/genesis_scripts/src/root-bash_profile:
```diff
@@ -1,14 +1,33 @@
+then
+	tail -n 10000 -f --pid="${GENESIS_PID}" "${GENESIS_LOGFILE}"
+else
+	echo "Something went wrong trying to tail the Genesis logfile:"
```
This should echo to `&2` to be cleanest, but this is super helpful!

byxorna commented on bootcd/rpms/genesis_scripts/src/run-genesis-bootloader:
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
```
Should drop a `set -e` here in case something blows up, so we bail early and the operator doesn't have to sift thru loads of failing output from subsequent commands

byxorna commented on bootcd/rpms/genesis_scripts/src/run-genesis-bootloader:
```diff
@@ -0,0 +1,24 @@
+if [[ ! -z $GENESIS_MODE ]]
+then
+    # disable task execution prompting when doing initial target run
+    GENESIS_PROMPT_TIMEOUT=0 /usr/bin/genesis-bootloader
```
Drop the full path from this binary. Without the full path you can dev easier by screwing with $PATH

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -25,45 +26,55 @@ Scripts and configuration files used by Genesis in the bootcd image
 
 %files
 %defattr(-, root, root)
+%config /etc/init.d/genesis
```
This isn't a config, just a file. Replacement​ behavior is different with configs in rpm

nshalman replied:
Most of the stuff in `%post` wasn't idempotent and would go wrong if you install the package more than once so I'm improving that as well.

byxorna replied:
Good, thank you! That will save much headache for devs

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -29,41 +30,55 @@ Scripts and configuration files used by Genesis in the bootcd image
-# TODO undo this hack
-cp  /etc/init/tty.conf.override /etc/init/tty.conf
-/usr/bin/patch /etc/sysconfig/init < /etc/sysconfig/init.diff
+sed '/.bash_profile.genesis_scripts/d' -i /root/.bash_profile
```
Is this necessary? Where is this initially inserted into the profile?

nshalman replied:
It was previously appended at the end. In an attempt to be idempotent I wanted to have a single line appended at the end of the file that I could easily delete and then re-append.

Your comment about `%config` annotations with regards to upgrades or re-installs made me think about what happens if the package gets installed again or upgraded. None of which actually happens in the build process so this is largely academic...

I don't know how to detect in `%post` if this is a fresh install or an upgrade.
I can switch to using `grep` as a gate instead of trying to be idempotent.

byxorna replied:
Scriptlets in rpm get the number of existing packages of this name on the system after the action completes. Use this to test in what context your scriptlet is executing and only perform profile inserts during new install 

From the fedora link below:

```
	install	upgrade	uninstall
 %pretrans	$1 == 0	$1 == 0	(N/A)
 %pre	$1 == 1	$1 == 2	(N/A)
 %post	$1 == 1	$1 == 2	(N/A)
 %preun	(N/A)	$1 == 1	$1 == 0
 %postun	(N/A)	$1 == 1	$1 == 0
 %posttrans	$1 == 0	$1 == 0	(N/A)

```

https://fedoraproject.org/wiki/Packaging:Scriptlets
https://www.redhat.com/archives/rpm-list/2002-August/msg00106.html

byxorna commented on bootcd/rpms/genesis_scripts/genesis_scripts.spec:
```diff
@@ -29,41 +30,55 @@ Scripts and configuration files used by Genesis in the bootcd image
-cp  /etc/init/tty.conf.override /etc/init/tty.conf
-/usr/bin/patch /etc/sysconfig/init < /etc/sysconfig/init.diff
+sed '/.bash_profile.genesis_scripts/d' -i /root/.bash_profile
+echo '. /root/.bash_profile.genesis_scripts' >> /root/.bash_profile
```
I'd gate this with a grep to not double insert on upgrade, or only perform this action if the argument to the %post indicates this is a new install, not upgrade

byxorna commented on bootcd/rpms/genesis_scripts/src/autologin:
```diff
@@ -1,3 +1,4 @@
 #! /bin/bash
 # /root/.bash_profile starts genesis-bootloader
+export HOME=/root
```
Is this going to effect all accounts, not just root?

nshalman replied:
The comment in that file is no longer accurate and should be cleaned out and replaced with a better one.

That file only gets used by `agetty` on the serial ports (not even on the graphical ttys which use `mingetty`) to log in as root. And to be clear, there aren't really any other usable accounts in a genesis boot environment:
```[root@genesis ~]# cat /etc/passwd
root:x:0:0:root:/root:/bin/bash
bin:x:1:1:bin:/bin:/sbin/nologin
daemon:x:2:2:daemon:/sbin:/sbin/nologin
adm:x:3:4:adm:/var/adm:/sbin/nologin
lp:x:4:7:lp:/var/spool/lpd:/sbin/nologin
sync:x:5:0:sync:/sbin:/bin/sync
shutdown:x:6:0:shutdown:/sbin:/sbin/shutdown
halt:x:7:0:halt:/sbin:/sbin/halt
mail:x:8:12:mail:/var/spool/mail:/sbin/nologin
uucp:x:10:14:uucp:/var/spool/uucp:/sbin/nologin
operator:x:11:0:operator:/root:/sbin/nologin
games:x:12:100:games:/usr/games:/sbin/nologin
gopher:x:13:30:gopher:/var/gopher:/sbin/nologin
ftp:x:14:50:FTP User:/var/ftp:/sbin/nologin
nobody:x:99:99:Nobody:/:/sbin/nologin
vcsa:x:69:69:virtual console memory owner:/dev:/sbin/nologin
sshd:x:74:74:Privilege-separated SSH:/var/empty/sshd:/sbin/nologin
ntp:x:38:38::/etc/ntp:/sbin/nologin```

nshalman commented:
I'll squash this all down to a single commit in the next day or two.

roymarantz commented on bootcd/rpms/genesis_scripts/src/root-bash_profile:
```diff
@@ -1,14 +1,43 @@
+		GENESIS_PID=$(cat "${GENESIS_PIDFILE}")
+		break
+	else
+		sleep .1
```
I thought that sleep(1) had a minimum of one second sleep time???

nshalman replied:
Per http://man7.org/linux/man-pages/man1/sleep.1.html the GNU coreutils version of sleep(1) can do smaller increments:

"Unlike most implementations that require NUMBER be an integer, here NUMBER may be an arbitrary floating point number."

roymarantz commented:
this is a great improvement over the current method.  Why have the autologin of any ttys though?  That was done as a trigger for starting genesis-bootloader, but since it is started by init that should not be needed.  I'm even sure I'd have root's bashrc tail the log, but instead echo the command that would do that.  i.e. I've gotten used to having to run a command after connecting, but as long as I could ^-C out of the tail it would be OK.  My practice is usually to only watch the output when I need to interact with the server.

byxorna commented:
@roymarantz operators use the default log tailing behavior to aid debugging and to quickly check in during induction without screwing with a keyboard

roymarantz commented:
Even though it isn't my preference or use case, I'm OK with the auto tailing, but am uneasy with autologined connections being able to write.  i.e. reading would be OK, but you should have to login before the terminal listens to you

byxorna commented:
@roymarantz I agree with your hesitation here. Ideally, we would have a clean way to echo the log output straight to the TTY without any kind of shady autologin or similar. I don't really have a better idea than what we have currently though... Unless we configured Genesis to log with a specific syslog facility (i.e. USER7) and configured rsyslog statically to ship all logs tagged thusly to `/dev/console` as well as the normal log location (instead of relying on output redirection from the invoking bootstrap process). Now that I write it out... It sounds a bit more robust, no? Cc @nshalman

nshalman commented:
The code has been squashed down. I'm unclear on how to handle the feedback from @roymarantz.
As far as I can tell I've pretty accurately replicated the existing behavior as perceived by an operator with the improvement that it does the right thing with the serial port when I pass in the right `console=` settings on the kernel command line.

If there is a desire to implement a new feature where the behavior is different from the current behavior that is enabled by the fact that the bootloader is now elegantly launched by init rather than hackishly on a single TTY, that seems like a reasonable request to make, but I'm not sure it should block this incremental improvement.

byxorna commented on bootcd/rpms/genesis_scripts/src/genesis.init:
```diff
@@ -0,0 +1,25 @@
+case "$1" in
+    start)
+	echo -n $"Starting genesis bootloader:  "
+	/usr/bin/run-genesis-bootloader &>/dev/null &
```
This doesn't accurately daemonize the process I think. We aren't closing stdin, and this process is still in the parent's process group (nohup/daemonize will double fork to escape the current pgroup). I think this will be ok when executed by init because the pgroup won't be different, but if a user ran the start script from a terminal, exiting the terminal will send a term signal to this backgrounded process (I think). Have you tested this behavior?

nshalman replied:
Simplest solution seems to me to just use `nohup`.
Adding it doesn't seem to break anything, but I haven't done extensive testing on running the init script from the command line.
I can't  think of a good reason someone would want to do that.
Either it ran correctly from `init` at boot time or you're doing manual debugging running `genesis` manually...

I've pushed an updated commit that uses nohup here.

byxorna commented:
@nshalman I agree, it sounds outside the scope of this PR. It would be really nice to have though, so we should open a new ticket to track the move to a more robust system for showing Genesis behavior on console. Perhaps suggestions could be collected into a GH issue for this to be discussed further? @roymarantz do you concur?

nshalman commented:
@byxorna, @roymarantz: Is there any more feedback that needs to be addressed before this can land?

byxorna commented:
@nshalman i dont have the magic wizarding powers to perform arcane merges, so we need to wait for @roymarantz to land.

roymarantz commented:
@nshalman sorry, but the more I think about it the more I can't have auto-login in my environment.  If there were some way that could be optional (ideally with the default being not enabled) I would merge this. 

I postulate that a boot option (visible in /proc/cmdline) would be one way to do it, alternatively a different boot image.  Neither seem real appealing though. @byxorna can you come up with another way to do it?

Another possibility is have the init.d file spawn some background processes that tail -f the /var/log file, one for each console/tty.  It shouldn't matter if any of the processes hang and I don't think they'll abort of the tty isn't connected.

nshalman commented:
@roymarantz you **currently** have auto-login on /dev/tty1 ([See here.](https://github.com/tumblr/genesis/blob/c7581eef9d8ac5c8448f41ab4ef6e96c134ca96f/bootcd/rpms/genesis_scripts/src/tty.conf.override#L13)) If the last task doesn't shut down or reboot, that tty gets dropped to a root shell.

I don't understand how this implementation is measurably worse in that regard.

**Further** improvement to change the default behavior to do something different will be simplified by the groundwork this change puts into place.

If I were adding auto-login where it wasn't already present I would understand your objection and would write further code before requesting a merge.

roymarantz commented:
Interesting.  I guess I'm "lucky" that it doesn't work in my environment.  Does it work in yours?

nshalman commented:
I thought about this, did some experiments, and finally hit upon a workable and relatively simple solution.
The automatic login and log tailing only happens if the kernel command line contains the string `GENESIS_AUTOTAIL`

Note that this definitely changes the default behavior of the Genesis live environment on `tty1`.

nshalman commented:
I'm again leaving these as separate commits for the moment for review, but they should be squashed down before being applied.

roymarantz commented:
Thanks @nshalman 👍 
Did you fill out a [CLA agreement](http://static.tumblr.com/zyubucd/GaTngbrpr/tumblr_corporate_contributor_license_agreement_v1__10-7-14.pdf)?  (I have trouble checking right now)

byxorna commented:
@gtorre @roymarantz can you guys independently build and verify functionality of this before merging? There is no automated test coverage, so we need to manually verify specific usecases work (also, documenting these usecases in a TESTING.md file would be super helpful)

nshalman commented:
Squashed. CLA is in-flight and should land within 48 hours (expected in less than 24, though.)

@byxorna do you want a TESTING.md to be part of this merge or can it be separate?

Also, github doesn't provide an interface for commenting on the commit message, but feedback on the current commit message if it could be clearer is also welcome.

nshalman commented:
I have been using the `Makefile` pasted below to do my development and testing which I am happy to check in if it would be of use to others.

There are four run targets that I use to run the 4 tests to verify the test matrix of whether or not the serial port is active and whether or not `GENESIS_AUTOTAIL` was specified:

```make
BUILDER=genesis-builder
CONTAINER=genesis
# Run QEMU connecting VM ttyS1 to stdio
QEMU=qemu-system-x86_64 --enable-kvm -smp 2 -m 2048 -kernel output/genesis-vmlinuz -initrd output/genesis-initrd.img -serial null -serial stdio -monitor null
BASE_ARGS=rootflags=loop root=live:/genesis.iso rootfstype=auto ro vga=791 rd_NO_LUKS rd_NO_MD rd_NO_DM GENESIS_MODE=util console=tty1

all:
	@echo Did you mean "make output"?

docker-image:
	sudo docker build -t $(BUILDER) .
	touch docker-image

output: docker-image
	mkdir $@
	sudo docker run --name $(CONTAINER) --privileged=true -v $(PWD)/$@:/output $(BUILDER)
	sudo chown -R $(USER) $@

.PHONY: run-serial run-serial-notail run-tty1 run-tty1-notail

run-serial: output
	# Note: ttyS1 must come last for the serial port to get agetty on it.
	@echo This test should result in automatic tailing of log on both tty1 and ttyS1
	$(QEMU) -append "$(BASE_ARGS) console=ttyS1 GENESIS_AUTOTAIL"

run-serial-notail: output
	# Note: ttyS1 must come last for the serial port to get agetty on it.
	@echo This test should result in login prompts on both tty1 and ttyS1
	$(QEMU) -append "$(BASE_ARGS) console=ttyS1"

run-tty1: output
	@echo This test should result in automatic tailing of log on tty1 only
	$(QEMU) -append "$(BASE_ARGS) GENESIS_AUTOTAIL"

run-tty1-notail: output
	@echo This test should result in login prompt on both tty1 only
	$(QEMU) -append "$(BASE_ARGS)"

clean:
	sudo losetup -a | cut -d: -f1 | xargs -r -n 1 sudo losetup -d
	sudo rm -rf output docker-image
	sudo docker rm $(CONTAINER) || true
	sudo docker rmi $(BUILDER) || true
```

nshalman commented:
@roymarantz the CLA has been sent.

roymarantz commented:
Great.  some final testing and I'll merge and cut a new release
Thanks again.
Roy

On Fri, Jun 30, 2017 at 9:04 AM, Nahum Shalman <notifications@github.com>
wrote:

> @roymarantz <https://github.com/roymarantz> the CLA has been sent.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tumblr/genesis/pull/67#issuecomment-312261616>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AF2LSm9lkzOp06Xiy-6oiEtXHi1R_2erks5sJPJHgaJpZM4NlkW1>
> .
>

byxorna commented:
@nshalman separate pr would be nice, to keep this scoped. Your makefile would be useful to have committed as well

nshalman commented:
I've opened #72 with the Makefile and a TESTING.md.

roymarantz commented:
I built an image and tested it on baremetal and it worked as expected.
